logbus/ship/ship.go

```diff
@@ -0,0 +1,118 @@
+package ship
+
+import (
+	"context"
+	"time"
+
+	pb "github.com/earthly/cloud-api/logstream"
+)
+
+type streamer interface {
+	StreamLogs(ctx context.Context, man *pb.RunManifest, ch <-chan *pb.Delta, verbose bool) error
+}
+
+type LogShipper struct {
+	cl      streamer
+	ch      chan *pb.Delta
+	man     *pb.RunManifest
+	err     error
+	cancel  context.CancelFunc
+	done    chan struct{}
+	verbose bool
+}
+
+func NewLogShipper(cl streamer, man *pb.RunManifest, verbose bool) *LogShipper {
+	return &LogShipper{
+		cl:      cl,
+		man:     man,
+		ch:      make(chan *pb.Delta),
+		done:    make(chan struct{}),
+		verbose: verbose,
+	}
+}
+
+func (l *LogShipper) Write(delta *pb.Delta) {
+	l.ch <- delta
+}
+
+func (l *LogShipper) Start(ctx context.Context) {
+	go func() {
+		ctx, l.cancel = context.WithCancel(ctx)
+		defer l.cancel()
+		out := bufferedDeltaChan(ctx, l.ch)
+		l.err = l.cl.StreamLogs(ctx, l.man, out, l.verbose)
+		l.done <- struct{}{}
+	}()
+}
+
+func (l *LogShipper) Close() {
+	close(l.ch)
+	// Graceful attempt to drain any in-flight logs then force-quit after delay.
+	t := time.NewTimer(10 * time.Second)
+	defer t.Stop()
+	select {
+	case <-t.C:
+		l.cancel()
+	case <-l.done:
+		return
+	}
+}
+
+func (l *LogShipper) Err() error {
```

Comment:
this feels like a race, since it depends on the thread started via `Start()` to complete.

Reply:
Good point. It should use a lock.